_source/_posts/2021-08-13-reactive-java.md

```diff
@@ -0,0 +1,791 @@
+
+```java
+@Test
+public void subscribeOnWithPublishOnTest() throws InterruptedException {
+    Flux.range(1, 3)
+            .map(v -> debug(v, "map1"))
+            .publishOn(Schedulers.parallel())
+            .map(v -> debug(v * 100, "map2"))
+            .subscribeOn(Schedulers.boundedElastic())
+            .subscribe(w -> debug(w,"subscribe"));
+
+    Thread.sleep(5000);
+}
+```
+
+The log should look like the following lines:
+
+```
+2021-07-29 08:40:30.583 - INFO [boundedElastic-1] - element "1" [map1]
+2021-07-29 08:40:30.584 - INFO [parallel-1] - element "100" [map2]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "100" [subscribe]
+2021-07-29 08:40:30.585 - INFO [boundedElastic-1] - element "2" [map1]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "200" [map2]
+2021-07-29 08:40:30.585 - INFO [parallel-1] - element "200" [subscribe]
+...
+```
+
+As you can see, the first `map` operator executes in the _boundedElastic-1_ thread, and the second `map` operator and the `subscribe` consumer execute in the _parallel-1_ thread. Notice that the `subscribeOn` operator is invoked after `publishOn`, but it affects the root of the chain and the operators preceding `publishOn` anyways.
+
+A probably simplified [marble diagam](https://projectreactor.io/docs/core/release/reference/#howtoReadMarbles) for the example above might look like:
+
+{% img blog/reactive-java/marble-diagram.png alt:"Marble Diagram" width:"800" %}{: .center-image }
+
+## Reactive Java Spring Services
+
+In Spring WebFlux, it is assumed applications don't block, so non-blocking servers use a small fixed-size thread pool to handle requests, named **event loop** workers.
+
+In an ideal reactive scenario, all the architecture components are non-blocking, so there is no need to worry about the event loop freezing up (**reactor meltdown**). But sometimes, you will have to deal with legacy blocking code, or blocking libraries.
+
+So now, let's experiment with Reactor Schedulers in a Reactive Java application. Create a Spring WebFlux service with Okta security. The service will expose an endpoint to return a random integer.
+
+The implementation will call Java `SecureRandom` blocking code. Start by downloading a Spring Boot Maven project using [Spring Initializr](https://start.spring.io/). You can do it with the following `HTTPie` line:
+
+```shell
+http -d https://start.spring.io/starter.zip type==maven-project \
+  language==java \
+  bootVersion==2.5.3 \
+  baseDir==reactive-service \
+  groupId==com.okta.developer.reactive \
+  artifactId==reactive-service \
+  name==reactive-service \
+  packageName==com.okta.developer.reactive \
+  javaVersion==11 \
+  dependencies==webflux,okta
+```
```

Comment:
type is not necessary since it's the default. You could probably remove java too. Fine to leave as is too.